File: aexample/src/main/scala/org/judal/examples/scala/jdbc/E04_CreateTablesFromJDOXML.scala
package org.judal.examples.scala.jdbc

import org.junit.Test

import org.scalatest.Suite

import org.judal.storage.DataSource
import org.judal.storage.relational.RelationalDataSource

import org.judal.metadata.TableDef
import org.judal.metadata.SchemaMetaData
import org.judal.metadata.bind.JdoXmlMetadata

import org.judal.jdbc.JDBCRelationalDataSource

import org.judal.examples.Resources.getResourceAsStream

import collection.JavaConverters._

import scala.collection.mutable.MutableList
import scala.collection.JavaConversions._

/**
 * Example of how to create tables read from a JDO XML file
 */
class E04_CreateTablesFromJDOXML extends Suite {

	@Test def demo() = {
		val dataSource = E02_CreateAdditionalDataSource.create()
		
		E04_CreateTablesFromJDOXML.createSchemaObjects(dataSource)

		E04_CreateTablesFromJDOXML.dropSchemaObjects(dataSource)
		
		E02_CreateAdditionalDataSource.close(dataSource)
	}	
	
}

object E04_CreateTablesFromJDOXML {

	def createSchemaObjects(dataSource: JDBCRelationalDataSource) = {

		val xmlMeta = new JdoXmlMetadata (dataSource)
		
		// Load table definitions from a JDO XML file
		val metadata = xmlMeta.readMetadata(getResourceAsStream("jdometadata.xml"))		

		val options = Map[String,Object](
		    DataSource.CATALOG -> "PUBLIC",
		    DataSource.SCHEMA -> "PUBLIC" ).asJava

		// Create tables in the database
		metadata.tables.foreach { tdef => dataSource.createTable(tdef, options) }
	}

	def dropSchemaObjects(dataSource: RelationalDataSource) = {
		
		val xmlMeta = new JdoXmlMetadata (dataSource)
		
		// Load table definitions from a JDO XML file
		val metadata = xmlMeta.readMetadata(getResourceAsStream("jdometadata.xml"))		
		
		// Drop tables in the database in reverse order than they were created
		metadata.tables.toList.reverse.foreach { tdef => dataSource.dropTable(tdef.getName, false) }

	}
  
}

File: aexample/src/main/java/org/judal/examples/java/jdbc/E04_CreateTablesFromJDOXML.java
package org.judal.examples.java.jdbc;

import java.util.Map;

import javax.jdo.JDOException;

import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedList;

import org.junit.Test;
import static org.junit.Assert.assertTrue;

import org.judal.storage.DataSource;
import org.judal.storage.relational.RelationalDataSource;
import org.judal.metadata.TableDef;
import org.judal.jdbc.JDBCRelationalDataSource;
import org.judal.metadata.SchemaMetaData;
import org.judal.metadata.bind.JdoXmlMetadata;

import static org.judal.examples.Resources.getResourceAsStream;

/**
 * Example of how to create tables read from a JDO XML file
 */
public class E04_CreateTablesFromJDOXML {

	@Test
	public void demo() throws Exception {
		JDBCRelationalDataSource dataSource = E02_CreateAdditionalDataSource.create();
		
		createSchemaObjects(dataSource);

		dropSchemaObjects(dataSource);
		
		E02_CreateAdditionalDataSource.close(dataSource);
	}	

	public static void createSchemaObjects(JDBCRelationalDataSource dataSource) throws JDOException, IOException {

		JdoXmlMetadata xmlMeta = new JdoXmlMetadata (dataSource);
		
		// Load table definitions from a JDO XML file
		SchemaMetaData metadata = xmlMeta.readMetadata(getResourceAsStream("jdometadata.xml"));		

		Map<String,Object> options = new HashMap<>();
		options.put(DataSource.CATALOG, "PUBLIC");
		options.put(DataSource.SCHEMA, "PUBLIC");
		
		// Create tables in the database
		for (TableDef tdef : metadata.tables())
			dataSource.createTable(tdef, options);
		
		assertTrue(dataSource.exists("student", "U"));
		assertTrue(dataSource.exists("course", "U"));
		assertTrue(dataSource.exists("student_x_course", "U"));
	}

	public static void dropSchemaObjects(RelationalDataSource dataSource) throws JDOException, IOException {
		
		JdoXmlMetadata xmlMeta = new JdoXmlMetadata (dataSource);
		
		// Load table definitions from a JDO XML file
		SchemaMetaData metadata = xmlMeta.readMetadata(getResourceAsStream("jdometadata.xml"));		
		
		// Reverse the table list
		LinkedList<String> tableNames = new LinkedList<>();
		for (TableDef tdef : metadata.tables())
			tableNames.addFirst(tdef.getName());
		
		// Drop tables in the database
		for (String tableName: tableNames)
			dataSource.dropTable(tableName, false);
	}
	
}
